frontend/src/pages/GraphPF/TrafficAnimation/TrafficRendererPF.tsx

```diff
@@ -0,0 +1,496 @@
+    const pointsOnEdge = travelDuration / this.launchTime;
+
+    // how many TrafficPoints we need to render (an int). We need to render at least one point
+    // for any edge with traffic, even if it is not always visible.
+    const renderedPointsOnEdge = Math.ceil(pointsOnEdge);
+
+    // time it takes for point to complete animation. This will be longer than then travel time
+    // when numPoints on edge < 1.0
+    const animationDuration = Math.max(travelDuration, this.launchTime);
+
+    // the percentage of the animationDuration for which the TrafficPoint is visible
+    const percentVisible = Math.min(100, Math.round((travelDuration / animationDuration) * 100));
+
+    // a slight randomization for the first point, so not everything launches at the same time
+    const initialDelay = Math.random() * (renderedPointsOnEdge - pointsOnEdge) * 1000;
+
+    const animationDurationSeconds = `${animationDuration / 1000}s`;
+    const renderer =
+      this.type === TrafficEdgeType.RPS
+        ? getTrafficPointRendererForRpsSuccess(edge, animationDurationSeconds, percentVisible)
+        : getTrafficPointRendererForTcp(edge, animationDurationSeconds, percentVisible);
+    const errorRenderer =
+      this.type === TrafficEdgeType.RPS
+        ? getTrafficPointRendererForRpsError(edge, animationDurationSeconds, percentVisible)
+        : undefined;
+
+    /* Debugging
+    console.log(
+      `renderedPoints=${renderedPointsOnEdge} pointsOnEdge=${pointsOnEdge.toFixed(2)} launchTime=${(
+        this.launchTime / 1000
+      ).toFixed(2)} travelDuration=${(travelDuration / 1000).toFixed(2)} animationDuration:${(
+        animationDuration / 1000
+      ).toFixed(2)} percentVisible=${percentVisible} initialDelay=${(initialDelay / 1000).toFixed(2)}`
+    );
+    */
+
+    const points: Array<React.SVGProps<SVGElement>> = [];
+    for (let i = 0; i < pointsOnEdge; ++i) {
+      const animationDelay = `${i * this.launchTime + initialDelay}ms`;
+      // If there is no mix of success and error points, just iterate infinitely
+      const isInfinite = this.errorRate === 0 || this.errorRate === 100;
+      // Otherwise, instruct the last Traffic point on the edge to renew the animation with new points
+      const renew = !isInfinite && i + 1 === renderedPointsOnEdge;
+      const isErrorPoint = errorRenderer && Math.random() <= this.errorRate;
+      const point = isErrorPoint
+        ? errorRenderer.render(edge, animationDelay, isInfinite, renew ? this.renewAnimation(edge) : undefined)
+        : renderer.render(edge, animationDelay, isInfinite, renew ? this.renewAnimation(edge) : undefined);
+      points.unshift(point);
+    }
+
+    return <>{points.map(p => p)}</>;
+  }
+
+  // renewAnimation performs an innocuous "set" to force the edge to re-render and generate new points
+  private renewAnimation(edge: Edge): React.AnimationEventHandler {
+    return _elem => {
+      window.setTimeout(() => {
```

Comment:
window is redundant here.
```suggestion
      setTimeout(() => {
```